rhp/v2/rhp.go

```diff
@@ -280,45 +303,116 @@ type (
-func RPCSectorRootsCost(settings HostSettings, n uint64) types.Currency {
-	return settings.BaseRPCPrice.
-		Add(settings.DownloadBandwidthPrice.Mul64(n * 32)).  // roots
-		Add(settings.DownloadBandwidthPrice.Mul64(128 * 32)) // proof
+type RPCCost struct {
+	Base       types.Currency
+	Storage    types.Currency
+	Ingress    types.Currency
+	Egress     types.Currency
+	Collateral types.Currency
+}
+
+func (c RPCCost) Add(o RPCCost) RPCCost {
+	return RPCCost{
+		Base:       c.Base.Add(o.Base),
+		Storage:    c.Storage.Add(o.Storage),
+		Ingress:    c.Ingress.Add(o.Ingress),
+		Egress:     c.Egress.Add(o.Egress),
+		Collateral: c.Collateral.Add(o.Collateral),
+	}
+}
+
+func (c RPCCost) Total() (cost, collateral types.Currency) {
+	return c.Base.Add(c.Storage).Add(c.Ingress).Add(c.Egress), c.Collateral
 }
 
-// RPCReadCost returns the price of a Read RPC.
-func RPCReadCost(settings HostSettings, sections []RPCReadRequestSection) types.Currency {
-	sectorAccessPrice := settings.SectorAccessPrice.Mul64(uint64(len(sections)))
+// RPCReadCost returns the cost of a Read RPC.
+func RPCReadCost(sections []RPCReadRequestSection, proof bool, settings HostSettings) (RPCCost, error) {
+	// validate the request sections and calculate the cost
 	var bandwidth uint64
 	for _, sec := range sections {
-		bandwidth += sec.Length
-		bandwidth += 2 * uint64(bits.Len64(LeavesPerSector)) * 32 // proof
+		switch {
+		case uint64(sec.Offset)+uint64(sec.Length) > SectorSize:
+			return RPCCost{}, ErrOffsetOutOfBounds
+		case sec.Length == 0:
+			return RPCCost{}, errors.New("length cannot be zero")
+		case proof && (sec.Offset%LeafSize != 0 || sec.Length%LeafSize != 0):
+			return RPCCost{}, errors.New("offset and length must be multiples of SegmentSize when requesting a Merkle proof")
+		}
+
+		bandwidth += uint64(sec.Length)
+		if proof {
+			start := sec.Offset / LeafSize
+			end := (sec.Offset + sec.Length) / LeafSize
+			proofSize := RangeProofSize(LeavesPerSector, start, end)
+			bandwidth += proofSize * 32
+		}
 	}
-	if bandwidth < minMessageSize {
-		bandwidth = minMessageSize
-	}
-	bandwidthPrice := settings.DownloadBandwidthPrice.Mul64(bandwidth)
-	return settings.BaseRPCPrice.Add(sectorAccessPrice).Add(bandwidthPrice)
+
+	return RPCCost{
+		Base:   settings.BaseRPCPrice.Add(settings.SectorAccessPrice.Mul64(uint64(len(sections)))),
+		Egress: settings.DownloadBandwidthPrice.Mul64(bandwidth),
+	}, nil
 }
 
-// RPCAppendCost returns the price and collateral of a Write RPC with a single
-// append operation.
-func RPCAppendCost(settings HostSettings, storageDuration uint64) (price, collateral types.Currency) {
-	price = settings.BaseRPCPrice.
-		Add(settings.StoragePrice.Mul64(SectorSize).Mul64(storageDuration)).
-		Add(settings.UploadBandwidthPrice.Mul64(SectorSize)).
-		Add(settings.DownloadBandwidthPrice.Mul64(128 * 32)) // proof
-	collateral = settings.Collateral.Mul64(SectorSize).Mul64(storageDuration)
-	// add some leeway to reduce chance of host rejecting
```

Comment:
@n8maninger `hostd` stopped doing it so I copied that behaviour, just wanted to draw attention to it to ensure it's safe to do so